finality-aleph/src/session_map.rs

```diff
@@ -281,76 +264,95 @@ where
+        let first_block = first_block_of_session(session_id, self.period);
+        self.authority_provider.authority_data(first_block)
     }
 
-    fn catch_up_boundaries(&self, period: SessionPeriod) -> (SessionId, SessionId) {
+    /// Puts current and next session authorities in the session map.
+    /// If previous authorities are still available in `AuthorityProvider`, also puts them in the session map.
+    async fn catch_up(&mut self) -> SessionId {
         let last_finalized = self.finality_notificator.last_finalized();
 
-        let current_session = session_id_from_block_num(last_finalized, period);
-        let starting_session = SessionId(current_session.0.saturating_sub(PRUNING_THRESHOLD));
+        let current_session = session_id_from_block_num(last_finalized, self.period);
+        let starting_session = SessionId(current_session.0.saturating_sub(PRUNING_THRESHOLD - 1));
 
-        (starting_session, current_session)
-    }
+        debug!(target: "aleph-session-updater",
+            "Last finalized is {:?}; Catching up with authorities starting from session {:?} up to next session {:?}",
+            last_finalized, starting_session.0, current_session.0 + 1
+        );
 
-    pub async fn run(mut self, period: SessionPeriod) {
-        let mut notifications = self.finality_notificator.notification_stream();
+        // lets catch up with previous sessions
+        for session in starting_session.0..current_session.0 {
+            let id = SessionId(session);
+            if let Some(authority_data) = self.authorities_for_session(id) {
+                self.session_map.update(id, authority_data).await;
+            } else {
+                debug!(target: "aleph-session-updater", "No authorities for session {:?} during catch-up. Most likely already pruned.", id.0)
+            }
+        }
 
-        let (starting_session, current_session) = self.catch_up_boundaries(period);
+        // lets catch up with previous session
+        match self.authorities_for_session(current_session) {
+            Some(current_authority_data) => {
+                self.session_map
+                    .update(current_session, current_authority_data)
+                    .await
+            }
+            None => panic!(
+                "Authorities for current session {:?} must be available from the beginning",
+                current_session.0
+            ),
+        };
 
-        // lets catch up
-        for session in starting_session.0..=current_session.0 {
-            self.update_session(SessionId(session), period).await;
-        }
+        self.handle_first_block_of_session(current_session).await;
 
-        let mut last_updated = current_session;
+        current_session
+    }
+
+    pub async fn run(mut self) {
+        let mut notifications = self.finality_notificator.notification_stream();
+        let mut last_updated = self.catch_up().await;
```

Comment:
I think technically you can catch up to a session while the current session advances, and you'll never get a notification, potentially leaving you without one of the authorities. I guess this should only happen on startup, limiting impact in practice, and as far as I can tell it was also the case before these changes.

I don't think you have to do anything about this, leaving this comment so you can let me know if I'm wrong or in case you actually see a simple way of fixing this. Otherwise feel free to resolve.

Reply:
Not really, the magic of unbounded channel :P We have open notifications channel before calling catch_up

Reply:
Ah, if that's a guarantee of the notification stream then all's good.